kernel/posix/semaphore.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <pthread.h>
+#include <errno.h>
+
+/**
+ * @brief destroy semaphore.
+ *
+ * FIXME: EBUSY is not taken care as of now
+ *
+ * see IEEE 1003.1
+ */
+int sem_destroy(sem_t *semaphore)
+{
+	if (semaphore == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	k_sem_reset(semaphore);
+
+	return 0;
+}
+
+/**
+ * @brief Get value of semaphore
+ *
+ * See IEEE 1003.1
+ */
+int sem_getvalue(sem_t *semaphore, int *value)
+{
+	*value = (int) k_sem_count_get(semaphore);
+
+	return 0;
+}
+/**
+ * @brief initialize semaphore
+ *
+ * While using this API, pass pshare value as 0 as feature
+ * is not supported by zephyr
+ *
+ * See IEEE 1003.1
+ */
+int sem_init(sem_t *semaphore, int pshared, unsigned int value)
+{
+	if (value == 0 || value > CONFIG_SEM_VALUE_MAX) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	/* pshared is not supported in zephyr. For invalid
+	 * values of pshare return error
+	 */
+	if (pshared != 0) {
+		errno = ENOSYS;
+		return -1;
+	}
+
+	k_sem_init(semaphore, value, CONFIG_SEM_VALUE_MAX);
+
+	return 0;
+}
+
+/**
+ * @brief unlock a semaphore
+ *
+ * See IEEE 1003.1
+ */
+int sem_post(sem_t *semaphore)
+{
+	if (semaphore == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	k_sem_give(semaphore);
+	return 0;
+}
+
+static int take_and_convert(sem_t *semaphore, s32_t timeout)
+{
+	int temp;
+
+	temp = k_sem_take(semaphore, timeout);
+
+	if (temp == -EBUSY) {
+		errno = EAGAIN;
+		return -1;
```

Comment:
can you optimize this function bit using a ret variable.
like int ret = 0;
and assign ret = -1 on error.
And "return ret;" at the end.